### PR TITLE
test: restore `write_buffer_errors_propagate`

### PR DESCRIPTION
**Pull request by crepererum**
This was removed in #2203 due to insufficient mocking capabilities.

### CODE REVIEW

File: server/src/lib.rs [hunk 1]
```diff
@@ -2234,6 +2239,44 @@ mod tests {
         assert!(matches!(err, Error::CannotCreateDatabase { .. }));
     }
 
+    #[tokio::test]
+    async fn write_buffer_errors_propagate() {
```
Comment (alamb) [inline server/src/lib.rs:2243]: 👍

File: write_buffer/src/config.rs [hunk 2]
```diff
@@ -42,17 +51,29 @@ impl WriteBufferConfigFactory {
...
     /// When mock with identical name is already registered.
     pub fn register_mock(&mut self, name: String, state: MockBufferSharedState) {
+        self.set_mock(name, Mock::Normal(state));
+    }
+
+    /// Registers new mock that always fail.
+    ///
+    /// # Panics
+    /// When mock with identical name is already registered.
+    pub fn register_alway_fail_mock(&mut self, name: String) {
```
Comment (alamb) [inline write_buffer/src/config.rs:61]: There is a typo here:

```suggestion
    pub fn register_always_fail_mock(&mut self, name: String) {
```

File: write_buffer/src/config.rs [hunk 3]
```diff
@@ -247,14 +282,113 @@ mod tests {
...
         let mut factory = WriteBufferConfigFactory::new();
 
         let state = MockBufferSharedState::empty_with_n_sequencers(1);
         let mock_name = "some_mock";
         factory.register_mock(mock_name.to_string(), state.clone());
         factory.register_mock(mock_name.to_string(), state);
     }
+
+    #[test]
+    #[should_panic(expected = "Mock with the name 'some_mock' already registered")]
```
Comment (alamb) [inline write_buffer/src/config.rs:364]: I wonder if these new three tests add much coverage? Is the code path for registering the different mock types different enough to warrant testing all 4 permutations?

It seems to me a single test would be sufficient.

**Review by alamb — APPROVED** [review]
Thank you

File: write_buffer/src/config.rs (hunk 3 above)
Comment (crepererum) [reply write_buffer/src/config.rs:364]: I can live w/ a single test as well, maybe I'm just overdoing it :D

File: write_buffer/src/config.rs (hunk 2 above)
Comment (crepererum) [reply write_buffer/src/config.rs]: fixed

File: write_buffer/src/config.rs (hunk 3 above)
Comment (crepererum) [reply write_buffer/src/config.rs]: changed to a single test

**Pull request merged**